File: 2025/src/rust/day_one.rs
// Copyright (C) 2025 Rémy Cases
// See LICENSE file for extended copyright information.
// This file is part of adventOfCode project from https://github.com/remyCases/adventOfCode.

use std::path::Path;
use std::env;
use std::io::{Error, ErrorKind};
use nom::*;
use nom::error::Error as NomError;
use aoc_utils::*;

#[derive(Debug)]
enum Direction {
    Left,
    Right,
}

const DIAL_MAX: i32 = 100;

fn read_file_and_compute(file_path: &Path, part: argparse::ArgPart) -> io::Result<()>
{
    let mut dial: i32 = 50;
    let mut dial_at_zero: i32 = 0;

    parse_compute!(
        file_path, part,
        sequence::pair(
            combinator::map_res(
                character::complete::one_of::<&str, &str, NomError<_>>("LR"),
                |c: char| match c { 
                    'L' => Ok(Direction::Left), 
                    'R' => Ok(Direction::Right),
                    _ => Err("Invalid direction"),
                }
            ),
            combinator::map_res(
                character::complete::digit1,
                str::parse,
            ),
        ), 
        |(d, u): (Direction, i32)| {
            match d {
                Direction::Left => dial = (dial + DIAL_MAX - u % DIAL_MAX) % DIAL_MAX,
                Direction::Right => dial = (dial + u % DIAL_MAX) % DIAL_MAX,
            }
            if dial == 0 { dial_at_zero += 1; }
        },
        |(d, u): (Direction, i32)| {
            dial_at_zero += u / DIAL_MAX;
            let increment: i32 = u % DIAL_MAX;
            
            if increment == 0 { return; }
            match d {
                Direction::Left => {
                    if dial <= increment && dial != 0 { dial_at_zero += 1; }
                    dial = (dial + DIAL_MAX - increment) % DIAL_MAX;
                },
                Direction::Right => {
                    if dial + increment >= DIAL_MAX { dial_at_zero += 1; }
                    dial = (dial + increment) % DIAL_MAX;
                },
            };
        }
    );

    println!("DIAL AT ZERO: {}", dial_at_zero);
    Ok(())
}

main!("2025", "input_day_one");

File: 2025/src/nim/dayOne.nim
# Copyright (C) 2025 Rémy Cases
# See LICENSE file for extended copyright information.
# This file is part of adventOfCode project from https://github.com/remyCases/adventOfCode.

import os, parseutils

const MAX_DIAL = 100

type
    Orientations = enum
        Left, Right

proc parseRotation(line: string, increment: var int, orientation: var Orientations, cycles: var int): bool =
    var o: char
    var i: int = parseChar(line, o, 0)

    case o:
    of 'L':
        orientation = Orientations.Left
    of 'R':
        orientation = Orientations.Right
    else:
        return false

    var val: int = 0
    i.inc parseInt(line, val, i)

    increment = val mod MAX_DIAL
    cycles = val div MAX_DIAL
    return true


proc readFileAndCompute(filename: string, option = '1') =
    var increment: int = 0
    var orientation: Orientations
    var cycles: int = 0
    var dial: int = 50
    var dial_at_zero: int = 0

    for line in filename.lines:
        if not parseRotation(line, increment, orientation, cycles):
            echo "Error parsing ", line
            return

        case option:
        of '1':
            if increment == 0: continue

            case orientation:
            of Orientations.Left:
                dial = (dial + MAX_DIAL - increment) mod MAX_DIAL
            of Orientations.Right:
                dial = (dial + increment) mod MAX_DIAL
            if dial == 0: dial_at_zero.inc
        of '2':
            dial_at_zero += cycles
            if increment == 0: continue
            case orientation:
            of Orientations.Left:
                if dial <= increment and dial != 0: dial_at_zero.inc
                dial = (dial + MAX_DIAL - increment) mod MAX_DIAL
            of Orientations.Right:
                if dial + increment >= MAX_DIAL: dial_at_zero.inc
                dial = (dial + increment) mod MAX_DIAL
        else:
            return

    echo "DIAL AT ZERO: ", dial_at_zero

proc main*(part: char) =
    const file = "2025/data/input_day_one"
    let filename = getCurrentDir() / file
    readFileAndCompute(filename, part)

when isMainModule:
    main('1')
